src/coreclr/jit/flowgraph.cpp

```diff
@@ -4543,3 +4543,250 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, ji
+
+    assert((comp->fgFirstBB->bbPreds == nullptr) && !comp->fgFirstBB->hasTryIndex());
+    assert(postOrder[count - 1] == comp->fgFirstBB);
+    comp->fgFirstBB->bbIDom = nullptr;
+
+    // First compute immediate dominators.
+    unsigned numIters = 0;
+    bool     changed  = true;
+    while (changed)
+    {
+        changed = false;
+
+        // In reverse post order, except for the entry block (count - 1 is entry BB).
+        for (unsigned i = count - 1; i > 0; i--)
+        {
+            unsigned    poNum = i - 1;
+            BasicBlock* block = postOrder[poNum];
+
+            // Intersect DOM, if computed, for all predecessors.
+            BasicBlock* bbIDom = nullptr;
+            for (FlowEdge* pred = comp->BlockDominancePreds(block); pred; pred = pred->getNextPredEdge())
+            {
+                BasicBlock* domPred = pred->getSourceBlock();
+                if (!dfs->Contains(domPred))
+                {
+                    continue; // Unreachable pred
+                }
+
+                if ((numIters <= 0) && (domPred->bbPostorderNum <= poNum))
+                {
+                    continue; // Pred not yet visited
+                }
+
+                if (bbIDom == nullptr)
+                {
+                    bbIDom = domPred;
+                }
+                else
+                {
+                    bbIDom = IntersectDom(bbIDom, domPred);
+                }
+            }
+
+            // Did we change the bbIDom value?  If so, we go around the outer loop again.
+            if (block->bbIDom != bbIDom)
+            {
+                changed       = true;
+                block->bbIDom = bbIDom;
+            }
+        }
+
+        numIters++;
+    }
+
+    // Now build dominator tree.
+    DomTreeNode* domTree = new (comp, CMK_DominatorMemory) DomTreeNode[count]{};
+
+    // TODO-Quirk: Build the dominator tree by iterating in normal post order
+    // (ending up with sibling links in RPO) instead of this. It will cause
+    // diffs because things like RBO and copy prop depend on the exact order of
+    // visiting blocks in the dominator tree.
+    for (BasicBlock* block : comp->Blocks(comp->fgFirstBB->Next()))
+    {
+        if (!dfs->Contains(block))
+        {
+            continue;
+        }
+
+        unsigned    poNum  = block->bbPostorderNum;
+        BasicBlock* parent = block->bbIDom;
+
+        domTree[poNum].nextSibling                 = domTree[parent->bbPostorderNum].firstChild;
+        domTree[parent->bbPostorderNum].firstChild = block;
+    }
+
+//// Skip the root since it has no siblings.
+// for (unsigned i = 0; i < count - 1; i++)
+//{
+//    BasicBlock* block  = postOrder[i];
+//    BasicBlock* parent = block->bbIDom;
+//    assert(dfs->Contains(block) && dfs->Contains(parent));
+
+//    domTree[i].nextSibling                     = domTree[parent->bbPostorderNum].firstChild;
+//    domTree[parent->bbPostorderNum].firstChild = block;
+//}
```

Comment:
I'll submit a PR right after this one that replaces the above quirk with this code, so I left it here. It seems like it has some nice improvements.